Detectors/TRD/base/src/LTUParam.cxx

```diff
@@ -0,0 +1,249 @@
+    dyMaxInt = mgDyMax;
+    dyMinInt = mgDyMin;
+  }
+}
+
+float LTUParam::getElongation(int det, int rob, int mcm, int ch) const
+{
+  // calculate the ratio of the distance to the primary vertex and the
+  // distance in x-direction for the given ADC channel
+
+  int layer = det % 6;
+
+  float elongation = TMath::Abs(getDist(det, rob, mcm, ch) * mgInvX[layer]);
+
+  // sanity check
+  if (elongation < 0.001) {
+    elongation = 1.;
+  }
+  return elongation;
+}
+
+void LTUParam::getCorrectionFactors(int det, int rob, int mcm, int ch,
+                                    unsigned int& cor0, unsigned int& cor1, float gain) const
+{
+  // calculate the gain correction factors for the given ADC channel
+  float Invgain = 1.0;
+  if (mPidGainCorr == true)
+    Invgain = 1 / gain;
+
+  if (mPidTracklengthCorr == true) {
+    cor0 = (unsigned int)((1.0 * mScaleQ0 * (1. / getElongation(det, rob, mcm, ch))) * Invgain);
+    cor1 = (unsigned int)((1.0 * mScaleQ1 * (1. / getElongation(det, rob, mcm, ch))) * Invgain);
+  } else {
+    cor0 = (unsigned int)(mScaleQ0 * Invgain);
+    cor1 = (unsigned int)(mScaleQ1 * Invgain);
+  }
+}
+
+int LTUParam::getNtimebins() const
+{
+  // return the number of timebins used
+
+  return mNtimebins;
+}
+
+float LTUParam::getX(int det, int /* rob */, int /* mcm */) const
+{
+  // return the distance to the beam axis in x-direction
+
+  int layer = det % 6;
+  return mgX[layer];
+}
+
+float LTUParam::getLocalY(int det, int rob, int mcm, int ch) const
+{
+  // get local y-position (r-phi) w.r.t. the chamber centre
+
+  int layer = det % 6;
+  // calculate the pad position as in the TRAP
+  float ypos = (-4 + 1 + (rob & 0x1) * 4 + (mcm & 0x3)) * 18 - ch - 0.5; // y position in bins of pad widths
+  return ypos * mgWidthPad[layer];
+}
+
+float LTUParam::getLocalZ(int det, int rob, int mcm) const
+{
+  // get local z-position w.r.t. to the chamber boundary
+
+  int stack = (det % 30) / 6;
+  int layer = det % 6;
+  int row = (rob / 2) * 4 + mcm / 4;
+
+  if (stack == 2) {
+    if (row == 0)
+      return (mgZrow[layer * 6 + stack] - 0.5 * mgLengthOuterPadC0);
+    else if (row == 11)
+      return (mgZrow[layer * 6 + stack] - 1.5 * mgLengthOuterPadC0 - (row - 1) * mgLengthInnerPadC0);
+    else
+      return (mgZrow[layer * 6 + stack] - mgLengthOuterPadC0 - (row - 0.5) * mgLengthInnerPadC0);
+  } else {
+    if (row == 0)
+      return (mgZrow[layer * 6 + stack] - 0.5 * mgLengthOuterPadC1[layer]);
+    else if (row == 15)
+      return (mgZrow[layer * 6 + stack] - 1.5 * mgLengthOuterPadC1[layer] - (row - 1) * mgLengthInnerPadC1[layer]);
+    else
+      return (mgZrow[layer * 6 + stack] - mgLengthOuterPadC1[layer] - (row - 0.5) * mgLengthInnerPadC1[layer]);
+  }
+}
+
+float LTUParam::getPerp(int det, int rob, int mcm, int ch) const
+{
+  // get transverse distance to the beam axis
+  float y;
+  float x;
+  x = getX(det, rob, mcm) * getX(det, rob, mcm);
```

Comment:
this looks like x^2 and below there is again x*x. Is this correct?

Reply:
ah I missed this comment, but also bumped into, in comment below. Yes its wrong.